feat(python): handle PyCapsule interface objects in write_deltalake

kylebarron opened the pull request:
# Description

Adds support for the [Arrow PyCapsule interface](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html). 

Since pyarrow is already a required dependency, this takes the minimal route of converting pycapsule interface objects into pyarrow objects. This requires pyarrow 15 or higher for the stream conversion (https://github.com/apache/arrow/issues/39217).

This doesn't modify the existing hard-coded support for pyarrow and pandas

# Related Issue(s)

- closes https://github.com/delta-io/delta-rs/issues/2376

# Documentation

ion-elgreco commented:
@kylebarron can you fix the linting issues?

Also wondering, how we should typehint this now, since an input can have the c_stream attribute or not

kylebarron commented:
> @kylebarron can you fix the linting issues? Then we can merge it

I'm pretty packed but I can try to find some time soon.

> Also wondering, how we should typehint this now, since an input can have the c_stream attribute or not

You can use these type hints: https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html#protocol-typehints

ion-elgreco commented:
@kylebarron ah nice, do you mind adding those typehints when you find the time

kylebarron commented:
I believe I fixed the lint and fixed the type hinting.

In the future, a more involved PR could remove pyarrow as a required dependency entirely by passing the C stream pycapsule directly to Rust (arrow-rs has an example of how to do that [here](https://github.com/apache/arrow-rs/blob/9be0eb530de85a68358dc31c146f71ccb97739ca/arrow/src/pyarrow.rs#L416-L425))

ion-elgreco commented on python/deltalake/writer.py:
```diff
@@ -70,6 +71,15 @@
+    https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html
+    """
+
+    def __arrow_c_stream__(self, requested_schema: object | None = None) -> object: ...
```
```suggestion
    def __arrow_c_stream__(self, requested_schema: Optional[object] = None) -> object: ...
```

ion-elgreco commented on python/deltalake/writer.py:
```diff
@@ -70,6 +71,15 @@
+    https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html
+    """
+
+    def __arrow_c_stream__(self, requested_schema: Optional[object] = None) -> object: ...
```
```suggestion
    def __arrow_c_stream__(
        self, requested_schema: Optional[object] = None
    ) -> object: ...
```

ion-elgreco commented:
@kylebarron that would be nice, we could potentially make it opt-in since it would then only be needed for reading

ion-elgreco reviewed (approved):
Thanks for adding this!
